clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java

```diff
@@ -289,49 +129,40 @@ public Struct parseResponse(short version, ByteBuffer buffer) {
     /** indicates whether the API is enabled for forwarding **/
     public final boolean forwardable;
 
-    public final Schema[] requestSchemas;
-    public final Schema[] responseSchemas;
     public final boolean requiresDelayedAllocation;
 
-    ApiKeys(int id, String name, Schema[] requestSchemas, Schema[] responseSchemas) {
-        this(id, name, false, requestSchemas, responseSchemas);
-    }
+    public final ApiMessageType messageType;
 
-    ApiKeys(int id, String name, boolean clusterAction, Schema[] requestSchemas, Schema[] responseSchemas) {
-        this(id, name, clusterAction, RecordBatch.MAGIC_VALUE_V0, requestSchemas, responseSchemas);
+    ApiKeys(ApiMessageType messageType) {
+        this(messageType, false);
     }
 
-    ApiKeys(int id, String name, Schema[] requestSchemas, Schema[] responseSchemas, boolean forwardable) {
-        this(id, name, false, RecordBatch.MAGIC_VALUE_V0, true, requestSchemas, responseSchemas, forwardable);
+    ApiKeys(ApiMessageType messageType, boolean clusterAction) {
+        this(messageType, clusterAction, RecordBatch.MAGIC_VALUE_V0, false);
     }
 
-    ApiKeys(int id, String name, boolean clusterAction, boolean isEnabled, Schema[] requestSchemas, Schema[] responseSchemas) {
-        this(id, name, clusterAction, RecordBatch.MAGIC_VALUE_V0, isEnabled, requestSchemas, responseSchemas, false);
+    ApiKeys(ApiMessageType messageType, boolean clusterAction, boolean forwardable) {
+        this(messageType, clusterAction, RecordBatch.MAGIC_VALUE_V0, forwardable);
     }
 
-    ApiKeys(int id, String name, boolean clusterAction, byte minRequiredInterBrokerMagic,
-            Schema[] requestSchemas, Schema[] responseSchemas) {
-        this(id, name, clusterAction, minRequiredInterBrokerMagic, true, requestSchemas, responseSchemas, false);
+    ApiKeys(ApiMessageType messageType, boolean clusterAction, byte minRequiredInterBrokerMagic, boolean forwardable) {
+        this(messageType, clusterAction, minRequiredInterBrokerMagic, forwardable, true);
     }
 
     ApiKeys(
-        int id,
-        String name,
+        ApiMessageType messageType,
         boolean clusterAction,
         byte minRequiredInterBrokerMagic,
-        boolean isEnabled,
-        Schema[] requestSchemas,
-        Schema[] responseSchemas,
-        boolean forwardable
+        boolean forwardable,
+        boolean isEnabled
     ) {
+        short id = messageType.apiKey();
```

Comment:
It seems to me this check should be moved to ```ApiMessageTypeGenerator``` as it is a generated code from json.

Reply:
We can, but it doesn't seem particularly beneficial. It's important to generate code that is specific to each data class. Generic code is best written without generation.

We can move this code to a base class for ApiMessageType potentially, particularly if we decide to remove ApiKeys.

Reply:
>  it doesn't seem particularly beneficial

It seems to me the benefit is the error happens from runtime/test_runtime to build time (generate message code).

Reply:
Oh, I see what you mean. You're suggesting to validate during generation, not to include the validation in the generated code.

Reply:
I took a look at this and not clear how to validate this in the generator. I added unit tests to `ApiMessageTypeTest` instead and I think that's good enough for now. What do you think?

Reply:
That alternative looks good to me :)